File: v2/clients/desktop/src/renderer/src/lib/timefmt.ts
/**
 * WhatsApp-style time labels — the TS mirror of agentd/clients/timefmt.py.
 * One rule set for every 'when' in the app:
 *   today          -> "14:32"      (list rows)  /  "Today"     (date separators)
 *   yesterday      -> "Yesterday"
 *   within 7 days  -> "Tuesday"
 *   same year      -> "5 June"
 *   older          -> "3 Apr 1996"
 */

function startOfDay(ms: number): number {
  const d = new Date(ms)
  d.setHours(0, 0, 0, 0)
  return d.getTime()
}

function dayDelta(ms: number, nowMs: number): number {
  return Math.round((startOfDay(nowMs) - startOfDay(ms)) / 86_400_000)
}

export function timeLabel(ms: number): string {
  const d = new Date(ms)
  return `${String(d.getHours()).padStart(2, '0')}:${String(d.getMinutes()).padStart(2, '0')}`
}

function dayOrDate(ms: number, nowMs: number, days: number): string {
  const d = new Date(ms)
  if (days === 1) return 'Yesterday'
  if (days < 7) return d.toLocaleDateString('en-GB', { weekday: 'long' })
  if (d.getFullYear() === new Date(nowMs).getFullYear()) {
    return d.toLocaleDateString('en-GB', { day: 'numeric', month: 'long' }) // "5 June"
  }
  return d.toLocaleDateString('en-GB', { day: 'numeric', month: 'short', year: 'numeric' }) // "3 Apr 1996"
}

/** Label for a LIST row (sessions sidebar): clock time today, else day/date. */
export function whenLabel(ms: number, nowMs = Date.now()): string {
  if (!ms) return ''
  const days = dayDelta(ms, nowMs)
  return days <= 0 ? timeLabel(ms) : dayOrDate(ms, nowMs, days)
}

/** Label for a DATE SEPARATOR between messages: 'Today' instead of a clock. */
export function dayLabel(ms: number, nowMs = Date.now()): string {
  if (!ms) return ''
  const days = dayDelta(ms, nowMs)
  return days <= 0 ? 'Today' : dayOrDate(ms, nowMs, days)
}

/** Same calendar day? — decides where date separators go. */
export function sameDay(a: number, b: number): boolean {
  return startOfDay(a) === startOfDay(b)
}

File: v2/agent_runtime/clients/timefmt.py
"""WhatsApp-style time labels, shared by the Python clients (terminal REPL + CLI).

One rule set, applied everywhere a 'when' is shown:
  today            -> "14:32"        (a clock time is the most useful 'today' label)
  yesterday        -> "Yesterday"
  within 7 days    -> "Tuesday"      (weekday name)
  same year        -> "5 June"
  older            -> "3 Apr 1996"

Pure functions of (epoch, now) so they're unit-testable without freezing clocks.
The desktop client mirrors these rules in renderer/src/lib/timefmt.ts.
"""

from __future__ import annotations

import time as _time
from datetime import datetime


def _day_delta(dt: datetime, now: datetime) -> int:
    return (now.date() - dt.date()).days


def whatsapp_when(epoch: float, now: float | None = None) -> str:
    """Label for a LIST row (sessions list): time today, else a relative day/date."""
    if not epoch:
        return ""
    dt = datetime.fromtimestamp(epoch)
    ref = datetime.fromtimestamp(now if now is not None else _time.time())
    days = _day_delta(dt, ref)
    if days <= 0:
        return dt.strftime("%H:%M")
    return _day_or_date(dt, ref, days)


def whatsapp_day(epoch: float, now: float | None = None) -> str:
    """Label for a DATE SEPARATOR between messages: 'Today' instead of a clock."""
    if not epoch:
        return ""
    dt = datetime.fromtimestamp(epoch)
    ref = datetime.fromtimestamp(now if now is not None else _time.time())
    days = _day_delta(dt, ref)
    if days <= 0:
        return "Today"
    return _day_or_date(dt, ref, days)


def _day_or_date(dt: datetime, ref: datetime, days: int) -> str:
    if days == 1:
        return "Yesterday"
    if days < 7:
        return dt.strftime("%A")                      # weekday name
    if dt.year == ref.year:
        return f"{dt.day} {dt.strftime('%B')}"        # "5 June"
    return f"{dt.day} {dt.strftime('%b')} {dt.year}"  # "3 Apr 1996"
